bitbybit/src/bit_size.rs

```diff
@@ -0,0 +1,47 @@
+use proc_macro2::TokenStream;
+use quote::quote;
+use quote::spanned::Spanned;
+
+const BAD_SIZE: &'static str = "The specified storage size is invalid, valid is in range 1..=64";
+
+/// Explicitly declared storage type for `bitenum` enums.
+pub(crate) struct Bits {
+    pub(crate) path: syn::Path,
+    pub(crate) size: usize,
+}
+
+impl Bits {
+    pub(crate) fn base_type(&self) -> syn::Result<syn::Ident> {
+        let span = self.path.__span();
+        let ident_str = match self.size {
+            1..=8 => "u8",
+            9..=16 => "u16",
+            17..=32 => "u32",
+            33..=64 => "u64",
+            _ => return Err(syn::Error::new(span, BAD_SIZE)),
+        };
+        Ok(syn::Ident::new(ident_str, span))
+    }
+    fn is_legacy(&self) -> bool {
+        let is_ident = self.path.get_ident().is_some();
+        let is_native_int = [8, 16, 32, 64].contains(&self.size);
+        !is_native_int && is_ident
+    }
+    pub(crate) fn qualified_path(&self) -> syn::Result<TokenStream> {
+        let (path, base, size) = (&self.path, self.base_type()?, self.size);
+
+        if self.is_legacy() {
+            Ok(quote!(arbitrary_int::UInt::<#base, #size>))
```

Comment:
The way it should be done is:

1. Have separates crate for the proc macro and the `bitbybit` crate.
2. re-export both the macros and `arbitrary_int` in `bitbybit`
3. Use `::bitbybit::arbitrary_int::…` instead of `arbitrary_int` in the macro generation code.
4. Ideally, add an additional attribute argument to `bitenum` that let end-users customize the `::bitbybit`. `serde` and `enumset` are crates that allow this.
5. Finally, fully qualify the std lib types used in the macro (so it would be `::core::result::Result` instead of `Result`)

Reply:
Why change the current system? Dependency too loosely specified?

Reply:
The current system makes the macro "unhygienic". As in: it pulls identifiers from some surrounding scope, which is pretty poor.

Specifically for `Result`, consider someone importing a custom `Result` type like `use syn::Result`. It would result in a compilation error, since now the macro would refer to `syn::Result` when using `Result`.

For `arbitrary_int`, the issue is that we have to add it as an explicit dependency. Which is pretty poor user experience. There is also problems about keeping the two versions in sync. If we re-imported it ourselves, it would shift the responsibility to us, not to our many users.

Reply:
Agree about Result! Fully qualifying it would certainly fix it.

For arbitrary-int, I'm less sure. You certainly have a point, but I think it would also be annoying for users to deal with having two namespaces of arbitrary-int, which likely are mostly the same.